### PR TITLE
pnpm.fetchDeps: multiple workspaces; better pnpmInstallFlags

### PR DESCRIPTION
**Pull request by pyrox0**
Adds support for installing multiple PNPM workspace packages' dependencies with the `pnpmWorkspaces` variable, and adds `pnpmInstallFlags` to other parts of PNPM configuration. Also adds appropriate documentation.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/tools/pnpm/fetch-deps/default.nix
Comment (gepbird) [inline pkgs/development/tools/pnpm/fetch-deps/default.nix:18]: Having both `pnpmWorkspace` and `pnpmWorkspaces` will probably cause some unnecessary refactors and nitpicks when reviewing. Can we deprecate the other (not necessarily in this PR)?
```suggestion
```

File: pkgs/development/tools/pnpm/fetch-deps/default.nix
Comment (gepbird) [inline pkgs/development/tools/pnpm/fetch-deps/default.nix:21]: Can you mention this with an example in the docs?

**Review by gepbird — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/pnpm/fetch-deps/default.nix
Comment (emilazy) [reply pkgs/development/tools/pnpm/fetch-deps/default.nix:18]: Note that the deadline for breaking changes is the 25th, so there’s about 25½ hours left to merge this PR if that option is chosen.

File: pkgs/development/tools/pnpm/fetch-deps/default.nix
Comment (pyrox0) [reply pkgs/development/tools/pnpm/fetch-deps/default.nix:20]: Added a mention of this.

**Comment by pyrox0** [comment]
Forgot to push the commits with the related package changes. I expect a good ofborg eval, can't check locally atm sorry.

File: pkgs/by-name/ba/bash-language-server/package.nix
Comment (gepbird) [inline pkgs/by-name/ba/bash-language-server/package.nix:29]: Same for `astro-language-server`
```suggestion
      pnpmWorkspaces
```

**Review by gepbird — CHANGES_REQUESTED** [review]
Ofborg failed because you forgot to change `pnpmWorkspace` in `pnpmDeps`.

Also would you mind splitting up your first commit? It contains many formatting changes that makes the diff a bit hard to read.

Edit: eval fails for some other reason too

File: pkgs/development/tools/pnpm/fetch-deps/default.nix
Comment (gepbird) [inline pkgs/development/tools/pnpm/fetch-deps/default.nix:39]: This was the other eval error: `error: attribute 'pnpmWorkspace' missing`
```suggestion
    lib.throwIf (builtins.hasAttr "pnpmWorkspace" args)
```

Comment (emilazy) [reply pkgs/development/tools/pnpm/fetch-deps/default.nix:39]: `args ? pnpmWorkspace`

**Comment by gepbird** [comment]
Aside from the those and some minor formatting issues that we don't have time for, this LGTM! Tested the 2 affected packages on x86_64-linux. I'm going to sleep and may not be available before this should be merged, I'd appreciate if someone could run a nixpkgs-review on this.

File: pkgs/development/tools/pnpm/fetch-deps/default.nix
Comment (pyrox0) [reply pkgs/development/tools/pnpm/fetch-deps/default.nix:39]: Fixed

File: pkgs/by-name/ba/bash-language-server/package.nix
Comment (pyrox0) [reply pkgs/by-name/ba/bash-language-server/package.nix:29]: Fixed both of these instances.

**Comment by pyrox0** [comment]
> Ofborg failed because you forgot to change `pnpmWorkspace` in `pnpmDeps`.
> 
> Also would you mind splitting up your first commit? It contains many formatting changes that makes the diff a bit hard to read.
> 
> Edit: eval fails for some other reason too

The formatting changes come from the `throwIf` call being added, should I do a commit with just that and then a reformat commit after?

**Comment by emilazy** [comment]
My wisdom is to do `assert lib.throwIf … true;`, which formats nicely and is a somewhat pleasing “statement”. Also good for asserts.

**Comment by pyrox0** [comment]
> My wisdom is to do `assert lib.throwIf … true;`, which formats nicely and is a somewhat pleasing “statement”. Also good for asserts.

Fixed, a sanity check would be nice though just to double check the code actually makes sense

**Review by gepbird — APPROVED** [review]
Thanks!

Result of `nixpkgs-review pr 350751` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>nixpkgs-manual</li>
  </ul>
</details>

**Comment by gepbird** [comment]
Sorry for the early approval, packages fail for me even though ofborg and nixpkgs-review passed.
For example when trying to build `astro-language-server` the assert triggers.

File: pkgs/development/tools/pnpm/fetch-deps/default.nix
Comment (gepbird) [inline pkgs/development/tools/pnpm/fetch-deps/default.nix:40]: We want to throw when a user DID give a `pnpmWorkspace`
```suggestion
    assert (lib.throwIf (args ? pnpmWorkspace)
```

File: doc/languages-frameworks/javascript.section.md
Comment (Scrumplex) [inline doc/languages-frameworks/javascript.section.md:449]: nixfmt and add missing `;` to `pnpmInstallFlags`

```suggestion
{
  pnpm,
}:

stdenv.mkDerivation (finalAttrs: {
  pname = "foo";
  version = "0-unstable-1980-01-01";

  src = ...;

  pnpmInstallFlags = [ "--shamefully-hoist" ];

  pnpmDeps = pnpm.fetchDeps {
    inherit (finalAttrs) pnpmInstallFlags;
  };
})

```

File: doc/languages-frameworks/javascript.section.md
Comment (gepbird) [inline doc/languages-frameworks/javascript.section.md:481]:
```suggestion
If you need to use a PNPM workspace for your project, then set `pnpmWorkspaces = [ "<workspace project name 1>" "<workspace project name 2>" ]`, etc, in your `pnpm.fetchDeps` call,
```

File: doc/languages-frameworks/javascript.section.md
Comment (gepbird) [inline doc/languages-frameworks/javascript.section.md:488]:
```suggestion
pnpmWorkspaces = [ "@astrojs/language-server" ];
```

**Review by gepbird — CHANGES_REQUESTED** [review]
If we are talking about formatting here are my nits I held back due to the time pressure. We have ~10 hours to merge this with the `pnpmWorkspace` deprecation.

**Comment by emilazy** [comment]
More like 13¾ hours, for what it’s worth. I agree that we should be okay for time (sorry that I can’t review properly myself, I have my own PRs to rush in before the freeze).

**Comment by pyrox0** [comment]
Fixed all noted issues.

File: pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh
Comment (Scrumplex) [inline pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh:28]: I am a little unsure about this. Shellcheck would definitely warn here, but I assume wrapping this with quotes would break the loop. If only we had structured attrs by default :sweat_smile:

**Review by Scrumplex — APPROVED** [review]
LGTM

File: pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh
Comment (pyrox0) [reply pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh:28]: just ran shellcheck over this and it doesn't warn. I'm pretty sure I could get away without setting IFS, but it's an additional layer of protection against anything weird happening in the script. IFS does include space by default though, so theoretically it doesn't have to be set, I set it anyways for safety since we don't want to break up args by newlines.

Comment (emilazy) [reply pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh:28]: I encourage looking at some of the PRs in https://github.com/NixOS/nixpkgs/issues/205690 to get this hook ready for a `__structuredAttrs` world where this will be an array. However, that should not block this.

Comment (pyrox0) [reply pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh:28]: I'll be honest, most of the linked PRs are going right over my head since I don't understand exactly what would need to be changed for `__structuredAttrs` to work properly(or at all)

Comment (emilazy) [reply pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh:28]: If you comment on that issue I’m sure people would be able to help out making the hook work with and without `__structuerdAttrs`, but I don’t think that needs doing for 24.11; just something that will need doing at some point. (I kind of hope we can default `__structuredAttrs` and `strictDeps` on for 25.05, but it’d be a big push.)

Comment (Scrumplex) [reply pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh:28]: I agree with @emilazy here. Let's focus on getting this merged first. `__structuredAttrs` can be handled in a future PR.

**Review by gepbird — APPROVED** [review]
Changes LGTM! `astro-language-server`, `bash-language-server` and another pnpm package like `heroic` builds.

I started nixpkgs-review (on a different machine) and I'm getting 52 rebuilds compared to the previous 1 rebuild.

File: pkgs/development/tools/pnpm/fetch-deps/default.nix
Comment (doronbehar) [inline pkgs/development/tools/pnpm/fetch-deps/default.nix]: Should be removed? Or Perhaps kept above?

File: pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh
Comment (doronbehar) [inline pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh:27]: Hmm maybe `exit 2` or something if `pnpmWorkspace` (singular, not plural) is used?

File: pkgs/development/tools/pnpm/fetch-deps/default.nix
Comment (doronbehar) [inline pkgs/development/tools/pnpm/fetch-deps/default.nix:76]: Hmm OK I guess it'd simply have been more correct to remove that in the previous commit.

File: pkgs/development/tools/pnpm/fetch-deps/default.nix
Comment (doronbehar) [inline pkgs/development/tools/pnpm/fetch-deps/default.nix:40]: Nit: I think it would look and feel better if you'd use something like:

```
assert assertMsg ("foo" == "bar") "foo is not bar, silly"; ""
```

**Review by doronbehar — COMMENTED** [review]
Looks good overall, small comments left. Haven't read other comments in the thread.

File: pkgs/development/tools/pnpm/fetch-deps/default.nix
Comment (gepbird) [reply pkgs/development/tools/pnpm/fetch-deps/default.nix:40]: Can you give a more specific example? We've tried `args.pnpmWorkspace != null` but it caused this issue: https://github.com/NixOS/nixpkgs/pull/350751#discussion_r1813860174

File: pkgs/development/tools/pnpm/fetch-deps/default.nix
Comment (pyrox0) [reply pkgs/development/tools/pnpm/fetch-deps/default.nix]: It is removed. Not sure where you're seeing that.

File: pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh
Comment (pyrox0) [reply pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh:27]: There's an assert to guard for it in the fetchdeps default.nix, so pnpmWorkspace should never reach the script.

**Comment by gepbird** [comment]
nixpkgs-review looks good, I won't wait out the electron build.

File: pkgs/development/tools/pnpm/fetch-deps/default.nix
Comment (doronbehar) [reply pkgs/development/tools/pnpm/fetch-deps/default.nix]: > It is removed. Not sure where you're seeing that.

I reviewed the PR commit by commit, and this `${installFlags}` usage was removed only in a later commit. It is a nit pick, so I wouldn't have commented about this if this would have been my only comment.

File: pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh
Comment (doronbehar) [reply pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh:27]: > There's an assert to guard for it in the fetchdeps default.nix, so pnpmWorkspace should never reach the script.

If I understand correctly, that assertion is only done for the `pnpmDeps` (fixed-output) derivation, and not the main derivation that uses these `pnpmDeps`.

Comment (pyrox0) [reply pkgs/development/tools/pnpm/fetch-deps/pnpm-config-hook.sh:27]: Fair enough, I'll add the check.

File: pkgs/development/tools/pnpm/fetch-deps/default.nix
Comment (emilazy) [reply pkgs/development/tools/pnpm/fetch-deps/default.nix:40]: FWIW even with `assertMsg` you need a `true` on the end. I’m not sure what the difference between `throwIf` and `assertMsg` is beyond a negation.

**Review by gepbird — APPROVED** [review]
Throws an eval time error when `pnpmWorkspace` is both in a package derivation and in `pnpm.fetchDeps`, and a build time error when `pnpmWorkspace` is only in the derivation, as expected. Those 3 packages build, LGTM.

**Review by Scrumplex — APPROVED** [review]

**Comment by GetPsyched** [comment]
This should have been flagged in open PRs that use `pnpmWorkspace`; #340035 was just merged and this will likely break in `master`. Raised #352830 to fix it.

**Comment by Scrumplex** [comment]
> This should have been flagged in open PRs that use `pnpmWorkspace`; #340035 was just merged and this will likely break in `master`. Raised #352830 to fix it.

We don't really have tooling to find PRs that utilize `pnpm.fetchDeps`/`pnpm.configHook` currently. We tried our best previously with a change to pnpm store hashes but even missed a few PRs there

**Comment by gepbird** [comment]
We really need a tool to search PR diffs. GitHub seems to only search in comments:

**Comment by doronbehar** [comment]
As for the surprising hash change you experienced, I just wanted to report that I checked:

```
nix build -Lf. {astro,bash}-language-server.pnpmDeps --rebuild --keep-going
```

And they didn't report a change of hash.. Not sure why that happened to you. It could be that in a manner unrelated to the workspace change that package's `pnpmDeps` is not very reproducible.

**Comment by gepbird** [comment]
@doronbehar `pnpmDeps` hashes were changed for most of the packages, see https://github.com/NixOS/nixpkgs/pull/349093#issuecomment-2421873637

**Pull request merged**
